### PR TITLE
Adding unit tests for handlers

### PR DESCRIPTION
**Pull request by crystalhsieh**
Providing examples for how to write tests for our handlers by writing unit tests for all our handlers.
Unit tests for: Not Found Handler, Echo Handler, Status Handler, Static Handler

We have decided to only test HoldingHandler in integration tests, since it's used for testing multithreading and contains an infinite loop.

### CODE REVIEW

File: src/echo_handler_test.cc [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+TEST_F(EchoHandlerTest, InitTest) {
+  EXPECT_EQ(status_, RequestHandler::Status::OK);
+}
+
+TEST_F(EchoHandlerTest, HandleProperRequestTest) {
+  Response response;
+  EXPECT_EQ(echo_handler_.HandleRequest(*request_, &response), RequestHandler::Status::OK);
+  EXPECT_EQ(response.GetResponseCode(), Response::OK);
+  EXPECT_EQ(response.ToString(), EXPECTED_RESPONSE);
+}
```
Comment (jasonsliu) [inline src/echo_handler_test.cc:33]: newline

File: src/not_found_handler_test.cc [hunk 2]
```diff
@@ -0,0 +1,33 @@
+#define TEST_NOT_FOUND_HANDLER
```
Comment (jasonsliu) [inline src/not_found_handler_test.cc]: What's this for?

File: src/echo_handler_test.cc [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+class EchoHandlerTest : public ::testing::Test {
+ protected:
+  EchoHandlerTest() {
+    status_ = echo_handler.Init(uri_prefix_, config_);
+    request_ = Request::Parse(TEST_BUFFER);
+  }
+
+  EchoHandler echo_handler_;
+  const std::string uri_prefix_;
+  const NginxConfig config_;
```
Comment (jasonsliu) [inline src/echo_handler_test.cc]: Looks like NginxConfig and uri_prefix_ are never really used later on. Just move it in to the constructor so they're not member variables. (This applies to some of the other files too)

File: src/static_handler_test.cc [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+TEST_F(StaticHandlerTest, NullResponse) {
+  EXPECT_EQ(static_handler.HandleRequest(*request, nullptr), RequestHandler::Status::ERROR);
+}
+
+TEST_F(StaticHandlerTest, HandleProperRequestTest) {
+  EXPECT_EQ(static_handler.HandleRequest(*request, &response), RequestHandler::Status::OK);
+  EXPECT_EQ(response.GetResponseCode(), Response::OK);
+  EXPECT_EQ(response.ToString(), EXPECTED_RESPONSE);
+}
```
Comment (jasonsliu) [inline src/static_handler_test.cc]: newline

File: src/status_handler_test.cc [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+
+TEST_F(StatusHandlerTest, InitTest) {
+  EXPECT_EQ(status, RequestHandler::Status::OK);
+}
+
+TEST_F(StatusHandlerTest, HandleProperRequestTest) {
+  EXPECT_EQ(status_handler.HandleRequest(*status_request, &status_response), RequestHandler::Status::OK);
+  EXPECT_EQ(status_response.GetResponseCode(), Response::OK);
+  EXPECT_EQ(status_response.ToString(), EXPECTED_STATUS_RESPONSE);
+}
```
Comment (jasonsliu) [inline src/status_handler_test.cc]: newline

File: src/status_handler_test.cc [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+  const std::string uri_prefix;
+  const NginxConfig config;
+  const std::string STATUS_BUFFER = "GET /status HTTP/1.1\r\nAccept-Language: en-us\r\n\r\nname=1";
+  const std::string EXPECTED_STATUS_RESPONSE = "HTTP/1.1 200 OK\r\nContent-Type: text/html\r\n\r\n<!doctype html><html><body><h1>Server Status</h1><h3>Handler Info</h3></br><h2>Total Requests: 1</h2><h5>Number of requests for \"/echo\"</h5><ul style=\"list-style-type:none\"><li>200: 1</li></ul></body></html>";
+  RequestHandler::Status status;
+  std::unique_ptr<Request> echo_request;
+  std::unique_ptr<Request> status_request;
+  Response echo_response;
+  Response status_response;
+  ServerStatus server_status;
```
Comment (jasonsliu) [inline src/status_handler_test.cc]: This is looking a little crazy for a unit test. Can we be using mocks?

File: src/echo_handler_test.cc (hunk 3 above)
Comment (crystalhsieh) [reply src/echo_handler_test.cc]: done

File: src/echo_handler_test.cc (hunk 1 above)
Comment (crystalhsieh) [reply src/echo_handler_test.cc:33]: done

File: src/not_found_handler_test.cc (hunk 2 above)
Comment (crystalhsieh) [reply src/not_found_handler_test.cc]: following coding style of defining the test before the actual code https://github.com/UCLA-CS130/spaceteam/blob/master/src/connection_test.cc

**Comment by crystalhsieh** [comment]
Talked offline and removed all #define TEST_[file] since it's not currently being used.

**Pull request merged**